README.md
# CENTIPEDE Tutorial

> CENTIPEDE fits a bayesian hierarchical mixture model to learn TF-specific
> distribution of experimental data on a particular cell-type for a set of
> candidate binding sites described by a motif.

This is a practical tutorial for running [CENTIPEDE] with [DNase-Seq] data. It
explains how to prepare the data and how to run the analysis. The goal is to
predict if a putative transcription factor binding site is actually bound or
not. For details about the statistical models underlying the methods, please
see ([Pique-Regi, et al. 2011][1]).

[CENTIPEDE]: http://centipede.uchicago.edu/ 
[DNase-Seq]: https://en.wikipedia.org/wiki/DNase-Seq
[1]: http://genome.cshlp.org/content/21/3/447

Read the tutorial online or download the PDF:

- <https://slowkow.github.io/CENTIPEDE.tutorial>
- [centipede-tutorial.pdf][2]

[2]: https://github.com/slowkow/CENTIPEDE.tutorial/raw/master/vignettes/centipede-tutorial.pdf

This repository has functions to ease the use of CENTIPEDE:

- `centipede_data()` converts data to the format required for CENTIPEDE.
- `parse_region()` parses a string like "chr1:123-456".
- `read_bedGraph()` reads a bedGraph file with 4 columns: chrom, start, end, score.
- `read_fimo()` reads a text file output by FIMO and selects sites that meet a
  significance threshold.

I also provide example data that you can use to follow the tutorial:

- `cen` is a list with two items:
  - `cen$mat` is a matrix of read-start counts for 3,337 genomic regions.
  - `cen$regions` is a dataframe describing those regions.
- `site_cons` is a vector with mean conservation scores for the 3,337 regions,
  computed across 100 vertebrates.

# Installation

Install CENTIPEDE by running this in your shell (not within an R session):

```bash
wget http://download.r-forge.r-project.org/src/contrib/CENTIPEDE_1.2.tar.gz
R CMD INSTALL CENTIPEDE_1.2.tar.gz
```

Next, install the tutorial package:

```r
# This command didn't work for me.
# install.packages("CENTIPEDE", repos="http://R-Forge.R-project.org")

install.packages("devtools")
devtools::install_github("slowkow/CENTIPEDE.tutorial")
```


NEWS.md

CENTIPEDE.tutorial 1.2.1
------------------------

FIXES

* Fix read count centering around the motif. Thanks to @danrlu for reporting
  the issue and fix in issue #9

* Fix `DESCRIPTION` to pass `R CMD check`

CHANGES

* Add `@importFrom` to `R/functions.R`

* Use `janitor::clean_names()` to make column names more consistent

CENTIPEDE.tutorial 1.2
----------------------

* Initial release.


R/data.R
#' A dataset ready to use with CENTIPEDE.
#' 
#' @format A list with two items:
#' \describe{
#'   \item{mat}{A numeric matrix with counts of read start sites.
#'    Each row corresponds to a genomic site flanked by 100 bp. A single row
#'    contains information about the positive strand and the negative strand.
#'    So, for example, this dataset uses a transcription factor motif of length
#'    13 bp and flanks of length 100 bp. Therefore, the matrix has
#'    2 * (100 + 13 + 100) = 426 columns.
#'   }
#'   \item{regions}{A dataframe.
#'    Each row corresponds to a genomic site, in the same order as the
#'    matrix of counts. This has information from a run of FIMO using the 13 bp
#'    motif for STAT4.
#'   }
#' }
"cen"

#' A vector of mean conservation scores for the PWM sites in "cen"
#' 
#' @format A vector that corresponds to the data in "cen". Each item in the
#' vector is the mean conservation score across 100 verterbrates computed
#' with phastCons at UCSC.
#' @source \url{http://hgdownload.cse.ucsc.edu/goldenpath/hg19/phastCons100way/}
"site_cons"

R/motif_plots.r
#' Fit a motif with CENTIPEDE and plot the profile and posterior distribution.
#'
#' @param motif A string of a specific motif
#' @param cen an object
#' @param score fimo score cutoff
#' @param p.value p.value cutoff
#' @return plot of profile and posterior distribution of a specific motif
#' @details
#'    cen is a list returned by \code{\link{centipede_data}}. It has a dataframe
#'    called \code{regions} with one row for each putative motif match.
#'    It also has a matrix called \code{mat} with one row for each match,
#'    and each column is a nucleotide position relative to the center of the
#'    motif match. The values in this matrix are the count of read start sites
#'    at that position.
fitAndPlot <- function(motif, cen, score = NA, pvalue = NA) {
  # Match the name of the motif.
  idx <- cen$regions$pattern.name == motif
  if (!is.na(pvalue)) {
    idx <- idx & cen$regions$p.value <= pvalue
  }
  if (!is.na(score)) {
    idx <- idx & cen$regions$score >= score
  }
  motif.tfs <- list(
    mat = cen$mat[idx, , drop = FALSE],
    regions = cen$regions[idx, ]
  )

  # Estimate the motif length.
  motif.length <- median(sapply(
    X = motif.tfs$regions$matched.sequence,
    FUN = function(x) nchar(as.character(x))
  ))

  # Fit the model.
  motif.fit <- CENTIPEDE::fitCentipede(
    Xlist = list(DNase = motif.tfs$mat),
    Y = as.matrix(data.frame(
      Intercept = rep(1, nrow(motif.tfs$mat))
    ))
  )

  # Plot image of cutsites ranked by CENTIPEDE posteriors.
  #imageCutSites(motif.tfs$mat[order(motif.fit$PostPr),][c(1:100, (dim(motif.tfs$mat)[1]-100):(dim(motif.tfs$mat)[1])),])
  #imageCutSitesCombined(motif.tfs$mat[order(motif.fit$PostPr),][c(1:100, (dim(motif.tfs$mat)[1]-100):(dim(motif.tfs$mat)[1])),])

  # Plot estimated footprint.
  #imageCutSitesCombined(motif.tfs$mat[order(motif.fit$PostPr),][motif.fit$PostPr >= 0.9,])
  CENTIPEDE::plotProfile(
    Profile = motif.fit$LambdaParList[[1]],
    Mlen = motif.length
  )
  hist(x = motif.fit$PostPr, main = motif,  xlab = "Posterior Probability")
  legend(
    "top",
    legend = c(
      paste("Total Sites:", length(motif.fit$PostPr)),
      paste("Sites >= 0.99:", sum(motif.fit$PostPr >= 0.99))
    )
  )
}

plot_all_motifs <- function(cen_dat, folder = "motif_plots", score = 0, p.value = 0) {
  tfs <- readRDS(cen_dat)

  motifs <- unique(tfs$regions$pattern.name)
  dir.create(folder)
  pattern.removed <- c()

  for (motif in motifs) {

    filename <- file.path(folder, sprintf("%s.pdf", motif))

    pdf(filename)
    par(mfrow = c(2,1))

    res <- try({
      fitAndPlot(motif, tfs, p.value = p.value, score = score)
    }, silent = TRUE)

    if (inherits(res, "try-error")) {
      dev.off()
      file.remove(filename)
      pattern.removed <- append(pattern.removed, motif)
      next
    }

    dev.off()
  }

  cat(
    "######################\n ", length(pattern.removed), " removed Motifs:\n"
  )
  for (i in pattern.removed) {
    print(i)
  }
}


R/functions.R
#' Parse a genomic region into parts.
#'
#' @param region A string like "chr1:123-456".
#' @return A list with elements "chrom", "start", and "end".
#' @export
parse_region <- function(region) {
  xs <- strsplit(region, ":")[[1]]
  chrom <- xs[1]
  xs <- as.numeric(strsplit(xs[2], "-")[[1]])
  return(list("chrom" = chrom, "start" = xs[1], "end" = xs[2]))
}

#' Read a bedGraph file with 4 columns: chrom, start, end, score
#'
#' @param filename The file to read. No column names.
#' @param ... Additional parameters passed to \code{\link[readr]{read_tsv}}
#' @return A GRanges object.
#' @export
#' @importFrom readr read_tsv
read_bedGraph <- function(filename, ...) {
  dat <- read_tsv(
    filename, col_names = FALSE, ...)
  colnames(dat) <- c("chrom", "start", "end", "score")
  makeGRangesFromDataFrame(
    df = dat,
    keep.extra.columns = TRUE,
    ignore.strand = TRUE,
    starts.in.df.are.0based = TRUE
  )
}

#' Read a text file output by FIMO and select sites that meet a significance
#' threshold.
#'
#' @param fimo_file A FIMO text file with PWM match sites.
#' @param pvalue Select FIMO matches with p-value less than this.
#' @return A dataframe with one row per site.
#' @details
#'    For sites represented multiple times, the one with the maximum
#'    score is selected.
#' @export
#' @importFrom readr read_tsv
#' @importFrom janitor clean_names 
read_fimo <- function(fimo_file, pvalue = 1e-4, ...) {
  # Read the PWM sites.
  sites <- clean_names(read_tsv(fimo_file, ...))

  # Discard poor matches.
  sites <- sites[sites$p_value < pvalue, , drop = FALSE]

  # For sites represented multiple times, select the one with the max score.
  sites$region <- paste(sites$sequence_name, sites$start, sites$stop)

  # Sort sites according to regions and descending score.
  sites <- sites[with(sites, order(sequence_name, start, stop, -score)), ]

  # Find duplicated regions and remove the ones with lower scores.
  dups <- duplicated(sites$region)
  if(sum(dups) > 0) {
    sites <- sites[-which(dups), ]
  }

  # Drop the region column.
  sites <- sites[, !colnames(sites) %in% c("region")]

  # Ensure we have some matches.
  if (nrow(sites) == 0) {
    stop(sprintf("No significant sites for '%s'", fimo_file))
  }

  return(sites)
}

#' Count readstarts at each nucleotide position in each BED region.
#'
#' @param bam_file A BAM file with mapped DNase-Seq data.
#' @param fimo_file A FIMO text file with PWM match sites.
#' @param pvalue Select FIMO matches with p-value less than this.
#' @return A list with two items "mat" and "regions".
#' @details
#'    The returend matrix "mat" has one row for each region in the BED
#'    file, and one column for each genomic position.
#'    The returned dataframe "regions" described each region with a p-value
#'    and q-value from FIMO.
#' @export
#' @importFrom GenomicRanges GRanges
#' @importFrom IRanges IRanges
#' @importFrom Rsamtools scanBam ScanBamParam
centipede_data <- function(
  bam_file, fimo_file, pvalue = 1e-4, flank_size = 100, ...
) {
  # Read the FIMO output file.
  sites <- read_fimo(fimo_file = fimo_file, pvalue = pvalue, ...)

  # Upstream flank, the center of PWM match, and downstream flank.
  sites$start <- sites$start - flank_size
  sites$stop <- sites$stop + flank_size

  # Index the BAM file if necessary.
  bam_index_file <- sprintf("%s.bai", bam_file)
  if (!file.exists(bam_index_file)) {
    message("Indexing the BAM file... this may take several minutes.")
    indexBam(bam_file, overwrite = FALSE)
  }

  # Extract reads that overlap the PWM sites.
  bam_list <- scanBam(
    file = bam_file,
    param = ScanBamParam(
      which = GRanges(
        seqnames = sites$sequence_name,
        ranges = IRanges(
          start = sites$start,
          end = sites$stop
        )
      ),
      what = c("strand", "pos", "qwidth")
    )
  )

  if (length(bam_list) == 0) {
    stop(sprintf("No reads fall in sites from '%s'\n", fimo_file))
  }

  # Convert the list of "chr:start-end" regions to a dataframe.
  regions <- lapply(names(bam_list), parse_region)
  regions <- data.frame(
    sequence_name = unlist(sapply(regions, function(x) x["chrom"])),
    start = as.numeric(unlist(sapply(regions, function(x) x["start"]))),
    stop = as.numeric(unlist(sapply(regions, function(x) x["end"])))
  )
  regions$index <- 1:nrow(regions)

  # Grab score, p_value, and q_value, but drop regions with no reads.
  regions <- merge(regions, sites)

  # Sort the regions by coordinate.
  regions <- regions[with(regions, order(sequence_name, start, stop)),]
  bam_list <- bam_list[regions$index]

  # Drop unused columns.
  regions <- regions[,!colnames(regions) %in% c("index")]

  if (nrow(regions) != length(bam_list)) {
    stop(sprintf("ERROR: %d regions and %d in bam_list. '%s'\n",
                 nrow(regions), length(bam_list), fimo_file))
  }

  mat <- do.call(rbind, lapply(seq(1, length(bam_list)), function(i) {

    # The reads overlapping this region.
    region <- parse_region(names(bam_list)[i])
    region_len <- abs(region$end - region$start) + 1

    # Exclude reads with start positions outside the region.
    item <- bam_list[[i]]

    # take care of negative reads starting at end position
    neg_shift <- item$qwidth * as.numeric(item$strand == "-")
    item$pos <- item$pos + neg_shift

    idx <- item$pos >= region$start & item$pos <= region$end
    if (sum(idx) == 0) {
      return(rep(0, 2 * region_len))
    }
    strand <- item$strand[idx]
    position <- item$pos[idx]

    # Create a row that represents the flanking region surrounding a site,
    # each column is a nucleotide position relative to the center of the motif
    # match.
    # The values in this matrix are the number of read start sites that occur
    # at that position.
    # Since we have two strands, the first half of the matrix represents the
    # positive strand and the second half represents the negative strand.
    is.neg <- as.numeric(strand == "-")
    j <- 1 + position - region$start + (region_len * is.neg)

    as.numeric(table(factor(j, levels = seq(1, 2 * region_len))))
  }))
  rownames(mat) <- names(bam_list)

  list(mat = mat, regions = regions)
}

# TODO: Create better functions for viewing CENTIPEDE results.
#
# imageReadStarts <- function(mat, site.width = 13) {
#   # Try using image() ...
#   image(t(mat), useRaster = TRUE, axes = FALSE)
#   len <- ncol(mat) / 2
#   axis(
#     side = 1,
#     at = c(
#       (len / 2 - site.width) / ncol(mat),
#       (len / 2 + site.width) / ncol(mat)
#     ),
#     labels = c("", "")
#   )
#   axis(
#     side = 1,
#     at = c(
#       (len + (len / 2 - site.width)) / ncol(mat),
#       (len + (len / 2 + site.width)) / ncol(mat)
#     ),
#     labels = c("", "")
#   )
#
#   # Also try using grid.raster() ...
#   img <- 1 - mat / max(mat)
#   w <- convertUnit(unit(ncol(img),"pt"), "in", value=TRUE)
#   h <- convertUnit(unit(nrow(img),"pt"), "in", value=TRUE)
#
#   # dev.new(width=w, height=h)
#
#   grid.raster(
#     image = img,
#     width = unit(1, "npc"),
#     height = unit(1, "npc")
#   )
#
#   v = dataViewport(xData = d$x, yData = d$y)
#   grid.points(d$x,d$y, default.units="native", vp = v,
#               gp=gpar(col="white"), pch=8)
# }
